File: lib/plugins/aws/invokeLocal/runtimeWrappers/invoke.py
import subprocess
import argparse
import json
import logging
import sys
import decimal
from time import strftime, time
from importlib import import_module

def decimal_serializer(o):
    if isinstance(o, decimal.Decimal):
        f = float(o)
        if f.is_integer():
          return int(f)
        return f
    
class FakeLambdaContext(object):
    def __init__(self, name='Fake', version='LATEST', timeout=6, **kwargs):
        self.name = name
        self.version = version
        self.created = time()
        self.timeout = timeout
        for key, value in kwargs.items():
            setattr(self, key, value)

    def get_remaining_time_in_millis(self):
        return int(max((self.timeout * 1000) - (int(round(time() * 1000)) - int(round(self.created * 1000))), 0))

    @property
    def function_name(self):
        return self.name

    @property
    def function_version(self):
        return self.version

    @property
    def invoked_function_arn(self):
        return 'arn:aws:lambda:serverless:' + self.name

    @property
    def memory_limit_in_mb(self):
        return '1024'

    @property
    def aws_request_id(self):
        return '1234567890'

    @property
    def log_group_name(self):
        return '/aws/lambda/' + self.name

    @property
    def log_stream_name(self):
        return strftime('%Y/%m/%d') +'/[$' + self.version + ']58419525dade4d17a495dceeeed44708'

    @property
    def log(self):
        return sys.stdout.write

logging.basicConfig()

parser = argparse.ArgumentParser(
    prog='invoke',
    description='Runs a Lambda entry point (handler) with an optional event',
)

parser.add_argument('handler_path',
                    help=('Path to the module containing the handler function,'
                          ' omitting ".py". IE: "path/to/module"'))

parser.add_argument('handler_name', help='Name of the handler function')

if __name__ == '__main__':
    args = parser.parse_args()

    # this is needed because you need to import from where you've executed sls
    sys.path.append('.')

    module = import_module(args.handler_path.replace('/', '.'))
    handler = getattr(module, args.handler_name)

    input = json.load(sys.stdin)
    if sys.platform != 'win32':
        try:
            if sys.platform != 'darwin':
                subprocess.check_call('tty', stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        except (OSError, subprocess.CalledProcessError):
            pass
        else:
            sys.stdin = open('/dev/tty')

    context = FakeLambdaContext(**input.get('context', {}))
    result = handler(input['event'], context)
    sys.stdout.write(json.dumps(result, default=decimal_serializer, indent=4))

File: packages/serverless/lib/plugins/aws/invoke-local/runtime-wrappers/invoke.rb
require 'json'
require 'securerandom'

class FakeLambdaContext
  attr_reader :function_name, :function_version, :aws_request_id, :log_stream_name, :memory_limit_in_mb,
    :invoked_function_arn, :log_group_name, :deadline_ms

  def initialize(function_name: 'Fake', version: 'LATEST', timeout: 6, **options)
    @function_name = function_name
    @function_version = version
    @memory_limit_in_mb = 1024
    @timeout = timeout

    # Allow overriding defaults
    options.each do |k,v|
      instance_variable_set("@#{k}", v)
    end

    @aws_request_id = SecureRandom.uuid
    @invoked_function_arn = "arn:aws:lambda:aws-region:acct-id:function:#{@function_name}"
    @log_group_name = "/aws/lambda/#{@function_name}"
    @log_stream_name = Time.now.strftime('%Y/%m/%d') +'/[$' + @function_version + ']58419525dade4d17a495dceeeed44708'

    @created_time = Time.now
    @deadline_ms = (@created_time + @timeout).to_i * 1000
  end

  def get_remaining_time_in_millis
    [@timeout * 1000 - ((Time.now - @created_time) * 1000).round, 0].max
  end

  def log(message)
    puts message
  end
end


def attach_tty
  unless Gem.win_platform? || $stdin.tty? || !File.exist?("/dev/tty")
    $stdin.reopen "/dev/tty", "a+"
  end
rescue
  puts "tty unavailable"
end

if __FILE__ == $0
  unless ARGV[0] && ARGV[1]
    puts "Usage: invoke.rb <handler_path> <handler_name>"
    exit 1
  end

  handler_path, handler_name = ARGV

  input = JSON.load($stdin) || {}

  require("./#{handler_path}")

  # handler name is either a global method or a static method in a class
  # my_method or MyModule::MyClass.my_method
  handler_method, handler_class = handler_name.split('.').reverse
  handler_class ||= "Kernel"

  attach_tty

  context = FakeLambdaContext.new(**input.fetch('context', {}).transform_keys(&:to_sym))
  result = Object.const_get(handler_class).send(handler_method, event: input['event'], context: context)

  puts result.to_json
end
